src/kernels.jl

```diff
@@ -0,0 +1,411 @@
+
+abstract type KNNKernel end
+
+sort_idxs(::KNNKernel) = false
+
+# Function that lists all defined `KNNKernels`
+list_kernels() = subtypes(KNNKernel) 
+
+####
+# User Defined Kernel (UDK)
+####
+
+_nothing(x) = nothing
+
+"""
+    UDK(;func::Function = x->nothing, sort::Bool=false)
+
+Wrap a user defined nearest neighbors weighting function `func` as a `KNNKernel`.
+ 
+# Keywords
+
+- `func` : user-defined nearest neighbors weighting function. The function 
+   should have the signature `func(dists_matrix)::Union{Nothing, <:AbstractMatrix}`, 
+   where `dists_matrix` is the k-nearest neighbors distances matrix and outputs a matrix 
+   of the same shape as `dists_matrix`.   
+- `sort` : if true sorts the `dists_matrix` so that in each row the gives 
+   k-nearest neighbors in acesending order.
+  
+"""
+struct UDK{T<:Function} <: KNNKernel
+   func::T
+   sort::Bool
+   UDK(func::F, sort::Bool) where {F<:Function} = new{F}(func, sort)
+end
+
+UDK(;func= _nothing, sort=false) = UDK(func, sort)
+sort_idxs(kernel::UDK) = kernel.sort
+
+function get_weights(kernel::UDK, dists_matrix)
+    func = kernel.func
+    nsamples, K = size(dists_matrix)
+    weights = func(dists_matrix)
+    weights === nothing && return nothing
+    check_onebased_indexing("weights", weights)
+    (nsamples, K) == size(weights) || throw(
+        DimensionMismatch(
+            "Expected a $((nsamples, K)) knn weights, got a $(size(weights)) knn "*
+            "weights matrix"
+        )
+    )
+    eltype(weights) <: AbstractFloat || throw(
+        error(
+            "The element type of knn weights matrix must be a subtype of `AbstractFloat`"
+        )
+    )
+    return weights
+end
+
+####
+# Uniform Kernel (Uniform)
+####
+
+"""
+    Uniform()
+
+assigns uniform weights to all k-nearest neighbors.
+
+see also:
+[`Inverse`](@ref), [`ISquared`](@ref), [`Zavreal`](@ref)     
+"""
+struct Uniform <: KNNKernel end
+
+get_weights(::Uniform, dists_matrix) = nothing
+
+
+####
+# Dudani
+####
+
+"""
+    Dudani()
+
+Assigns the closest neighbor a weight of `1`, the furthest neighbor weight `0` and the 
+others are scaled between by a linear mapping.
+
+For more information see the paper by Geler et.al [Comparison of different weighting 
+schemes for the kNN classifier on time-series data](https://perun.pmf.uns.ac.rs/
+radovanovic/publications/2016-kais-knn-weighting.pdf).
+
+see also:
+[`Macleod`](@ref), [`DualU`](@ref), [`DualD`](@ref)
+"""
+struct Dudani <: KNNKernel end
+
+sort_idxs(::Dudani) = true
+
+function _dudani_weights(dists_matrix, inds)
+    i, j = inds.I
```

Comment:
Problem here:

```julia
julia> d=Dudani()
Dudani()

julia> MLJNearestNeighborsInterface.get_weights(d, rand(2,3))
ERROR: type Int64 has no field I
Stacktrace:
 [1] getproperty(::Int64, ::Symbol) at ./Base.jl:33
 [2] _dudani_weights(::Array{Float64,2}, ::Int64) at /Users/anthony/Dropbox/Julia7/MLJ/MLJNearestNeighborsInterface/src/kernels.jl:98
 [3] _broadcast_getindex_evalf at ./broadcast.jl:648 [inlined]
 [4] _broadcast_getindex at ./broadcast.jl:621 [inlined]
 [5] getindex at ./broadcast.jl:575 [inlined]
 [6] copy at ./broadcast.jl:876 [inlined]
 [7] materialize at ./broadcast.jl:837 [inlined]
 [8] get_weights(::Dudani, ::Array{Float64,2}) at /Users/anthony/Dropbox/Julia7/MLJ/MLJNearestNeighborsInterface/src/kernels.jl:107
 [9] top-level scope at REPL[13]:1
```

Reply:
Okay, never mind. I see this is only intended for your NeighDistsMatrix wrapper.